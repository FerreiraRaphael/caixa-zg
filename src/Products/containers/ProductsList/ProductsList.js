import React from 'react';

import List from '../../components/List';

class ProductsList extends React.Component {
  state = {
    cursor: 0
  };

  componentDidMount() {
    const { items } = this.props;
    if (items.length === 0) {
      this.props.fetchProducts();
    }
  }

  handleKeyPress = e => {
    const { cursor, result } = this.state;
    const { items } = this.props;
    // arrow up/down button should select next/previous list element
    if (e.key === `ArrowUp` && cursor > 0) {
      this.setState(prevState => ({
        cursor: prevState.cursor - 1
      }));
    } else if (e.key === `ArrowDown` && cursor < items.length - 1) {
      this.setState(prevState => ({
        cursor: prevState.cursor + 1
      }));
    }
  };

  render() {
    const { items, addItem, removeItem, deleteItem } = this.props;
    return (
      <List
        onKeyPress={this.handleKeyPress}
        selectedIndex={this.state.cursor}
        items={items}
        onAdd={addItem}
        onRemove={removeItem}
        onDelete={deleteItem}
      />
    );
  }
}

export default ProductsList;
