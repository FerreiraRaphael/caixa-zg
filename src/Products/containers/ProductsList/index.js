import { connect } from 'react-redux';

import { findIndexByProp } from '../../../lib/utility/array';
import ProductsList from './ProductsList';
import {
  addItem,
  removeItem,
  deleteItem
} from '../../../Purchases/modules/purchases';
import { fetchProducts } from '../../modules/products';

export const mapStateToProps = state => {
  const items = state.products.items.map(item => {
    const index = findIndexByProp('product', item.product)(
      state.purchases.items
    );
    const purchase = state.purchases.items[index];
    return { ...item, quantity: purchase ? purchase.quantity : 0 };
  });
  return {
    items
  };
};

export const mapDispatchToProps = dispatch => ({
  fetchProducts: () => {
    dispatch(fetchProducts());
  },
  addItem: product => {
    dispatch(addItem(product));
  },
  removeItem: product => {
    dispatch(removeItem(product));
  },
  deleteItem: product => {
    dispatch(deleteItem(product));
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(ProductsList);
