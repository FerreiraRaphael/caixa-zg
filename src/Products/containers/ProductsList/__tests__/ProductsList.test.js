import React from 'react';
import { shallow, mount } from 'enzyme';

import ProductsList from '../ProductsList';
import { mapStateToProps } from '../index';

describe('ProductsList Component', () => {
  const setup = (propsOverride = {}, mounting = false) => {
    const props = {
      items: [],
      fetchProducts: jest.fn(),
      ...propsOverride
    };

    const mountFunction = mounting ? mount : shallow;

    const wrapper = mountFunction(<ProductsList {...props} />);

    return {
      props: wrapper.props(),
      wrapper
    };
  };

  it('should render', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should call fetchProducts on component mount', () => {
    const { props } = setup({ items: [] }, true);
    expect(props.fetchProducts).toBeCalled();
  });

  describe('mapStateToProps', () => {
    it('should return correct props', () => {
      const state = {
        purchases: {
          items: [
            {
              name: 'Amaciante',
              sku: 'AMA',
              quantity: 10,
              product: '5a69bbbe085eb50006b7515b',
              price: 12,
              subtotal: 120,
              discount: 0,
              total: 120
            },
            {
              name: 'Sabão',
              sku: 'SAB',
              quantity: 6,
              product: '5a69bbd2085eb50006b75162',
              price: 0.8,
              subtotal: 4.8,
              discount: 0,
              total: 4.8
            }
          ]
        },
        products: {
          items: [
            {
              name: 'Sabão',
              sku: 'SAB',
              product: '5a69bbd2085eb50006b75162',
              price: 0.8
            },
            {
              name: 'Produto',
              sku: 'PRO',
              product: '1',
              price: 0.8
            },
            {
              name: 'Amaciante',
              sku: 'AMA',
              product: '5a69bbbe085eb50006b7515b',
              price: 12
            }
          ]
        }
      };
      const props = mapStateToProps(state);
      expect(props).toEqual({
        items: [
          {
            name: 'Sabão',
            sku: 'SAB',
            quantity: 6,
            product: '5a69bbd2085eb50006b75162',
            price: 0.8
          },
          {
            name: 'Produto',
            sku: 'PRO',
            quantity: 0,
            product: '1',
            price: 0.8
          },
          {
            name: 'Amaciante',
            sku: 'AMA',
            quantity: 10,
            product: '5a69bbbe085eb50006b7515b',
            price: 12
          }
        ]
      });
    });
  });
});
