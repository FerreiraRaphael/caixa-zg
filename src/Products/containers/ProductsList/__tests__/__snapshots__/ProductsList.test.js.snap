// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsList Component should render 1`] = `
<List
  items={Array []}
  onKeyPress={[Function]}
  selectedIndex={0}
/>
`;
